File: android/app/src/main/java/com/vxin/app/data/repository/ContactRepository.kt
package com.vxin.app.data.repository

import com.vxin.app.data.api.ContactApi
import com.vxin.app.data.api.MessageApi
import com.vxin.app.data.model.Contact
import com.vxin.app.data.model.CreateGroupBody
import com.vxin.app.data.model.CreatePrivateBody
import com.vxin.app.data.model.FriendRequest
import com.vxin.app.data.model.FriendRequestBody
import com.vxin.app.data.model.HandleRequestBody
import com.vxin.app.data.model.SearchUser
import com.vxin.app.data.model.SendRequestResponse
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ContactRepository @Inject constructor(
    private val contactApi: ContactApi,
    private val messageApi: MessageApi,
    socketManager: com.vxin.app.core.realtime.SocketManager,
) {
    /** 好友申请相关实时事件（新申请/被通过） */
    val friendEvents = socketManager.friendEvents
    /** 联系人在线/离线 */
    val presenceEvents = socketManager.presenceEvents

    suspend fun contacts(): List<Contact> = contactApi.contacts()

    suspend fun search(q: String): List<SearchUser> = contactApi.search(q)

    suspend fun sendFriendRequest(toId: String, message: String): SendRequestResponse =
        contactApi.sendRequest(FriendRequestBody(toId, message))

    suspend fun receivedRequests(): List<FriendRequest> = contactApi.receivedRequests()

    suspend fun sentRequests() = contactApi.sentRequests()

    suspend fun handleRequest(id: String, accept: Boolean) =
        contactApi.handleRequest(id, HandleRequestBody(if (accept) "accept" else "reject"))

    suspend fun deleteContact(id: String) = contactApi.deleteContact(id)

    suspend fun setRemark(id: String, remark: String) =
        contactApi.setRemark(id, com.vxin.app.data.model.RemarkBody(remark))

    suspend fun block(id: String) = contactApi.block(id)

    suspend fun unblock(id: String) = contactApi.unblock(id)

    suspend fun blocked() = contactApi.blocked()

    /** 创建/获取私聊会话，返回 conversationId */
    suspend fun createPrivate(userId: String): String =
        messageApi.createPrivate(CreatePrivateBody(userId)).conversationId

    /** 创建群聊，返回 conversationId */
    suspend fun createGroup(name: String, memberIds: List<String>): String =
        messageApi.createGroup(CreateGroupBody(name, memberIds)).conversationId
}

File: ios/Vxin/Data/Repositories/ContactRepository.swift
import Foundation

struct FriendRequestBody: Encodable { let toId: String; let message: String }
struct HandleRequestBody: Encodable { let action: String }
struct SendRequestResponse: Decodable { let success: Bool?; let autoAccepted: Bool? }
struct CreatePrivateBody: Encodable { let userId: String }
struct CreateGroupBody: Encodable { let name: String; let memberIds: [String] }
struct CreateConversationResponse: Decodable { let conversationId: String; let groupNumber: String? }
private struct RemarkBody: Encodable { let remark: String }

struct BlockedUser: Decodable, Identifiable {
    let id: String
    var username: String = ""
    var avatar: String = ""
    enum CodingKeys: String, CodingKey { case id, username, avatar }
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(String.self, forKey: .id)
        username = (try? c.decode(String.self, forKey: .username)) ?? ""
        avatar = (try? c.decode(String.self, forKey: .avatar)) ?? ""
    }
}

/// 联系人/好友/会话创建。与 Android ContactRepository 等价。
final class ContactRepository {
    static let shared = ContactRepository()
    private init() {}

    private let api = APIClient.shared

    func contacts() async throws -> [Contact] {
        try await api.send("api/users/contacts")
    }

    func search(_ q: String) async throws -> [SearchUser] {
        let encoded = q.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? q
        return try await api.send("api/users/search?q=\(encoded)")
    }

    func sendFriendRequest(toId: String, message: String = "") async throws -> SendRequestResponse {
        try await api.send("api/users/friend-request", method: "POST", body: FriendRequestBody(toId: toId, message: message))
    }

    func receivedRequests() async throws -> [FriendRequest] {
        try await api.send("api/users/friend-requests")
    }

    func handleRequest(id: String, accept: Bool) async throws {
        let _: EmptyResponse = try await api.send(
            "api/users/friend-request/\(id)/handle", method: "POST",
            body: HandleRequestBody(action: accept ? "accept" : "reject")
        )
    }

    /// 创建/获取私聊会话，返回 conversationId
    func createPrivate(userId: String) async throws -> String {
        let res: CreateConversationResponse = try await api.send(
            "api/messages/conversation/private", method: "POST", body: CreatePrivateBody(userId: userId)
        )
        return res.conversationId
    }

    /// 创建群聊，返回 conversationId
    func createGroup(name: String, memberIds: [String]) async throws -> String {
        let res: CreateConversationResponse = try await api.send(
            "api/messages/conversation/group", method: "POST", body: CreateGroupBody(name: name, memberIds: memberIds)
        )
        return res.conversationId
    }

    // ── 好友管理：删除/备注/拉黑 ──
    func deleteContact(_ id: String) async throws {
        let _: EmptyResponse = try await api.send("api/users/contacts/\(id)", method: "DELETE")
    }

    func setRemark(_ id: String, remark: String) async throws {
        let _: EmptyResponse = try await api.send("api/users/contacts/\(id)/remark", method: "PUT", body: RemarkBody(remark: remark))
    }

    func block(_ id: String) async throws {
        let _: EmptyResponse = try await api.send("api/users/block/\(id)", method: "POST")
    }

    func unblock(_ id: String) async throws {
        let _: EmptyResponse = try await api.send("api/users/block/\(id)", method: "DELETE")
    }

    func listBlocked() async throws -> [BlockedUser] {
        try await api.send("api/users/me/blocked")
    }
}
